File: crates/mathhook-python/examples/basic_usage.py
#!/usr/bin/env python3
"""
MathHook Python Basic Usage Examples

This example demonstrates the fundamental operations available
in the MathHook Python bindings.

Last Updated: 2025-12-28T1200
"""

from mathhook import (
    symbol, symbols, parse,
    sin, cos, tan, exp, log, sqrt, abs_expr,
    pprint, init_printing
)

def main():
    print("=" * 60)
    print("MathHook Python - Basic Usage Examples")
    print("=" * 60)

    # =========================================================================
    # Symbol Creation
    # =========================================================================
    print("\n1. Symbol Creation")
    print("-" * 40)

    # Single symbol
    x = symbol('x')
    print(f"Single symbol: x = {x}")

    # Multiple symbols
    a, b, c = symbols('a b c')
    print(f"Multiple symbols: a={a}, b={b}, c={c}")

    # =========================================================================
    # Expression Creation with Operators
    # =========================================================================
    print("\n2. Expression Creation (Operator Overloading)")
    print("-" * 40)

    # Arithmetic operations
    expr1 = x + 2
    print(f"x + 2 = {expr1}")

    expr2 = x - 3
    print(f"x - 3 = {expr2}")

    expr3 = x * 5
    print(f"x * 5 = {expr3}")

    expr4 = x / 2
    print(f"x / 2 = {expr4}")

    expr5 = x ** 2
    print(f"x ** 2 = {expr5}")

    # Complex expression: quadratic
    quadratic = a * x**2 + b * x + c
    print(f"Quadratic: a*x² + b*x + c = {quadratic}")

    # =========================================================================
    # Mathematical Functions
    # =========================================================================
    print("\n3. Mathematical Functions")
    print("-" * 40)

    print(f"sin(x) = {sin(x)}")
    print(f"cos(x) = {cos(x)}")
    print(f"tan(x) = {tan(x)}")
    print(f"exp(x) = {exp(x)}")
    print(f"log(x) = {log(x)}")
    print(f"sqrt(x) = {sqrt(x)}")
    print(f"abs(x) = {abs_expr(x)}")

    # =========================================================================
    # Simplification
    # =========================================================================
    print("\n4. Simplification")
    print("-" * 40)

    # Combine like terms
    expr = x + x + x
    print(f"x + x + x = {expr.simplify()}")

    # Arithmetic simplification
    expr = parse("2 + 3 + 5")
    print(f"2 + 3 + 5 = {expr.simplify()}")

    # Polynomial simplification
    expr = x**2 + 2*x + 1 + x**2 - 1
    print(f"x² + 2x + 1 + x² - 1 = {expr.simplify()}")

    # =========================================================================
    # Parsing Expressions
    # =========================================================================
    print("\n5. Parsing Expressions")
    print("-" * 40)

    expr1 = parse("x^2 + 2*x + 1")
    print(f"Parsed 'x^2 + 2*x + 1': {expr1}")

    expr2 = parse("sin(x) + cos(y)")
    print(f"Parsed 'sin(x) + cos(y)': {expr2}")

    expr3 = parse("2x + 3y")  # Implicit multiplication
    print(f"Parsed '2x + 3y': {expr3}")

    # =========================================================================
    # Expansion and Factoring
    # =========================================================================
    print("\n6. Expansion and Factoring")
    print("-" * 40)

    # Expand (x + 1)^2
    expr = (x + 1) ** 2
    expanded = expr.expand()
    print(f"(x + 1)² expanded = {expanded}")

    # Factor x^2 - 1
    expr = x**2 - 1
    factored = expr.factor()
    print(f"x² - 1 factored = {factored}")

    # =========================================================================
    # Output Formats
    # =========================================================================
    print("\n7. Output Formats")
    print("-" * 40)

    expr = x**2 + 2*x + 1

    # String representation
    print(f"String: {expr}")

    # LaTeX format
    latex = expr.to_latex()
    print(f"LaTeX: {latex}")

    # Pretty printing
    print("Pretty print:")
    pprint(expr)

    # =========================================================================
    # Summary
    # =========================================================================
    print("\n" + "=" * 60)
    print("Basic usage examples completed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: crates/mathhook-node/examples/basic-usage.ts
#!/usr/bin/env tsx

/**
 * Basic MathHook Node.js/TypeScript Usage Examples
 *
 * This example demonstrates the fundamental operations available
 * in the MathHook Node.js bindings using TypeScript.
 */

import {
  Expression,
  symbols,
  parse,
  SmartEquationSolver,
} from "../index.js";

console.log("MathHook TypeScript Basic Usage Examples\n");

// ===== Expression Creation =====
console.log("Creating Mathematical Expressions:");

// Create symbols and constants
const [x, y] = symbols("x y");
const two = Expression.integer(2);
const three = Expression.integer(3);
const five = Expression.integer(5);

console.log(`Variable x: ${x.format()}`);
console.log(`Variable y: ${y.format()}`);
console.log(`Constant 2: ${two.format()}`);
console.log(`Constant 3: ${three.format()}\n`);

// ===== Basic Arithmetic Operations =====
console.log("Basic Arithmetic Operations:");

// Addition
const sum = Expression.add([x, two]);
console.log(`x + 2 = ${sum.format()}`);

// Multiplication
const product = Expression.mul([x, three]);
console.log(`x * 3 = ${product.format()}`);

// Exponentiation
const power = Expression.pow(x, two);
console.log(`x^2 = ${power.format()}`);

// Chained operations
const complex = Expression.add([Expression.mul([two, x]), three]);
console.log(`2x + 3 = ${complex.format()}\n`);

// ===== Expression Simplification =====
console.log("Expression Simplification:");

// Simple arithmetic simplification
const arithmetic = Expression.add([two, three]);
console.log(`Before: 2 + 3 = ${arithmetic.format()}`);
console.log(`After:  ${arithmetic.simplify().format()}`);

// Algebraic simplification
const algebraic = Expression.add([x, x, Expression.mul([two, three])]);
console.log(`Before: x + x + 2*3 = ${algebraic.format()}`);
console.log(`After:  ${algebraic.simplify().format()}\n`);

// ===== Equation Creation and Solving =====
console.log("Equation Solving:");

try {
  const solver = new SmartEquationSolver();

  // Simple equation: x = 5
  const equation1 = Expression.equation(x, five);
  console.log(`Equation 1: ${equation1.format()}`);
  const [solution1, explanation1] = solver.solveWithEquation(equation1, x.asSymbol()!);
  console.log(`Solution 1 valid: ${solution1.isValidSolution()}, count: ${solution1.solutionCount()}`);

  // Linear equation: 2x + 3 = 7
  const seven = Expression.integer(7);
  const leftSide = Expression.add([Expression.mul([two, x]), three]);
  const equation2 = Expression.equation(leftSide, seven);
  console.log(`Equation 2: ${equation2.format()}`);
  const [solution2, explanation2] = solver.solveWithEquation(equation2, x.asSymbol()!);
  console.log(`Solution 2 valid: ${solution2.isValidSolution()}, count: ${solution2.solutionCount()}\n`);
} catch (error: unknown) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(`Solver error: ${message}\n`);
}

// ===== Integrated Parsing =====
console.log("Integrated Mathematical Expression Parsing:");

try {
  // Direct parsing with automatic language detection
  const parsed1 = parse("2*x + sin(y)");
  console.log(`Auto-detect "2*x + sin(y)": ${parsed1.format()}`);
  console.log(`Simplified: ${parsed1.simplify().format()}`);

  // LaTeX automatic detection
  const parsed2 = parse("\\frac{x}{2} + y^2");
  console.log(`LaTeX "\\frac{x}{2} + y^2": ${parsed2.format()}`);

  // Simple expression
  const parsed3 = parse("sin(x) + cos(y)");
  console.log(`"sin(x) + cos(y)": ${parsed3.format()}`);

  // Format conversion
  const expr = parse("x^2");
  console.log(`Expression: ${expr.format()}\n`);
} catch (error: unknown) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(`Parser error: ${message}\n`);
}

// ===== Polynomial Operations =====
console.log("Polynomial Operations:");

// Create a quadratic polynomial: x^2 + 2x + 1
const quadratic = Expression.add([
  Expression.pow(x, two),
  Expression.mul([two, x]),
  Expression.integer(1),
]);

console.log(`Quadratic: ${quadratic.format()}`);
console.log(`Simplified: ${quadratic.simplify().format()}`);

// Create a more complex expression: (x + y)^2
const binomial = Expression.pow(Expression.add([x, y]), two);
console.log(`Binomial: ${binomial.format()}`);
console.log(`Simplified: ${binomial.simplify().format()}\n`);

console.log("All basic examples completed successfully!");
console.log("Try running the advanced examples next: tsx advanced-usage.ts");
